### PR TITLE
Type Promotion and Special Number Test Cases

### PR DESCRIPTION
**Pull request by rdspring1**
Test infrastructure for #1187

### CODE REVIEW

File: test/test_jit_cuda_fuser.py [hunk 1]
```diff
@@ -721,7 +723,45 @@ def test_binary_ops(self):
...
             if (dtype not in self.int_types) or (op not in skip_for_integer):
-                self._binary_test_helper(op, dtype)
+                self._binary_test_helper(op, dtype, True)  # random data
+                # disabled special numbers because of incorrect handling
+                # self._binary_test_helper(op, dtype, False) # special numbers
+
+    @unittest.skipIf(not RUN_CUDA, "requires CUDA")
+    @unittest.skipIf(GRAPH_EXECUTOR != ProfilingMode.PROFILING,
+                     "Requires fusion optimization pass to be effective")
+    def test_binary_ops_type_promotion(self):
```
Comment (shmsong) [inline test/test_jit_cuda_fuser.py:733]: Just curious what kind of errors were happening due to mixed data type. The might be more failing cases in mixed value types, value types meaning `N-Dim Tensor (N>0)`, `0-Dim Tensor`, `Scalar`.

The type promotion rule on eager mode treats inputs of different value types differently, e.g:
`Tensor + Scalar` and `3dim Tensor + 0-dim Tensor`. `test_category_rule` had some simple coverage on these scenarios.

**Review by shmsong — APPROVED** [review]
Overall looks good to me and really appreciate improving type and op tests. 👍 

Had one minor question on type promotion test. 

Approving for now but would recommend seeing @jjsjann123's feedback before merging.

File: test/test_jit_cuda_fuser.py (hunk 1 above)
Comment (rdspring1) [reply test/test_jit_cuda_fuser.py:733]: The errors are similar to this issue https://github.com/csarofeen/pytorch/issues/1167

File: test/test_jit_cuda_fuser.py [hunk 2]
```diff
@@ -721,7 +723,45 @@ def test_binary_ops(self):
...
+        data_types = [
+            torch.int32,
+            torch.int64,
+            # torch.float16,
+            torch.float32,
+            torch.float64
+        ]
+        # disabled bf16 data type - Issue #1165
```
Comment (jjsjann123) [inline test/test_jit_cuda_fuser.py]: nitpick: #1165 -> #1167 ?!

File: test/test_jit_cuda_fuser.py [hunk 3]
```diff
@@ -441,57 +441,7 @@ def t(x: torch.Tensor, y: torch.Tensor, z: torch.Tensor):
...
-            return o
-        t_jit = torch.jit.script(t)
-        x = torch.randn(4, 8, 32, 32, dtype=torch.float, device="cuda")
-        jit_o = t_jit(x, 2.0)
-        jit_o = t_jit(x, 2.0)
-        o = t(x, 2.0)
-        self.assertEqual(o, jit_o)
-        self.assertGraphContains(t_jit.graph_for(x, 2.0), FUSION_GUARD)
-
-    @unittest.skipIf(not RUN_CUDA, "requires CUDA")
```
Comment (jjsjann123) [inline test/test_jit_cuda_fuser.py:457]: Can we keep this test and put a TODO to remove them in #1187 ?

I'm slightly nervous with deleting this, while some ops in the replacement tests being disabled due to bfloat failures.

**Review by jjsjann123 — CHANGES_REQUESTED** [review]
LGTM, minor comment on keeping the existing `test_unary_ops` until we reach the same op coverage in replacement tests.

File: test/test_jit_cuda_fuser.py (hunk 2 above)
Comment (rdspring1) [reply test/test_jit_cuda_fuser.py]: It should be #1185

**Comment by rdspring1** [comment]
`test_unary_ops` is still a stand-alone test from `test_unary_ops_type_promotion`.
https://github.com/csarofeen/pytorch/blob/89e6a89c257f1b9c74dca78865f024b639693344/test/test_jit_cuda_fuser.py#L482-L515
https://github.com/csarofeen/pytorch/blob/89e6a89c257f1b9c74dca78865f024b639693344/test/test_jit_cuda_fuser.py#L520-L565

**Review by jjsjann123 — APPROVED** [review]
sorry I was dumb earlier :stuck_out_tongue_closed_eyes:

**Comment by csarofeen** [comment]
Not dumb :-( simply missed it.

**Pull request merged**